### DISCUSSION TITLE
Programmatic Profile Switching

### DISCUSSION THREAD

**Discussion by noisyneil** [elgatosf/elgato#45] [Help]
2025 UPDATE:

Thanks to @roycetech for making me aware that AppleScript is much faster than it used to be for profile switching. The macro described below now uses AppleScript to achieve the same result and is _much_ easier to set up and use than the old applet method!

Hi, 

TL;DR: Is programmatic profile switching feasible?

Many SD users on the Keyboard Maestro forum are desperate to get some kind of programmatic profile switching going. We've been emailing Elgato for a couple of years now, and there's been no real movement. 

This is a shame, because if you could automatically switch profiles depending on what you're doing in a particular app (for example), this would exponentially multiply the usefulness and immediacy of the SD.

As an example, I've come up with a (janky) way to switch profiles automatically for different Logic Pro plugins. It's works, but it's incredibly complicated and doesn't feel seamless:

[HERE IS THE MACRO](https://forum.keyboardmaestro.com/t/triggering-stream-deck-profiles-for-logic-pro-plugins/23357)

If this simply isn't possible, fair enough, but something as simple as switching profiles externally seems like it _might_ be possible, which is why I'm asking this here. Methods I can envision:

Switching via MIDI
Shell script
Javascript
Applescript

Any help, or a summary of the state of play, would be hugely appreciated.

Thanks for your time!

**Comment by UweSchwo** [comment]
Yes, programmatic profile switching would be a great thing. May be at least we get a more flexible Windows detection, so I could have different profiles for different open documents or dialog boxes by title evaluation. Having only the exe file started as trigger is not enough.

**Comment by AdamDobay** [comment]
Hi @noisyneil, as it happens I'm in a very similar conversation with Elgato myself! I'm advocating for the Stream Deck as an ADHD accessibility tool (lowering cognitive barriers of entry and counteracting working memory issues through shortcut mapping and macros), and I've been in conversation with Support around specifically whether the Stream Deck would be able to profile switch based on what browser tab / web app is open.

Interestingly, this is something that I actually see hampering the Stream Deck's usability as more and more software solutions are rolled out as web-based or web app-based with no development going into standalone PC/Mac executable versions.

In the first round of support I was recommended to find a plugin developer who might want to implement this feature through the Stream Deck SDK, but I talked to a plugin dev who confirmed that it's a core functionality of the Windows Stream Deck that it can identify .exe files only and switch profiles based on that. When I took this info back to support, they escalated my case, so I'm currently waiting on their reply.

I'm happy to keep you updated as I hear back from them.

What I'm wondering right now is if what you've devised for Mac might be implemented using similar logic but adapted to Windows? I'm way in over my head on this topic so right now I'm wondering if it's even possible.

Let me know what you think!

**Reply by noisyneil** [reply]
All you need to do is open an application associated with an SD profile. Most of the complexity in the macro above is to do with _when_ this happens because Logic Pro plugin windows are finicky.

Here's a more straightforward example of automation logic that I'm sure you should be able to achieve with AutoHotkey:

- Trigger on browser front window change
- If front window URL contains "github.com", open githubtrigger.exe

**Comment by UweSchwo** [comment]
Maybe it's easier if we suggest concrete ways of implementation. Here my suggestions:
1. Activate a profile by command line,  e. g.: StreamDeck.exe /d:StreamDeck1 /p:MyAccountingKeys
2. Let the user assign key combinations to profiles, e.g: MyAccountingKeys: CTRTL+ALT+F9
3. Activate a profile by a part of current window title in addition to the .exe, e.g. "my accounting.xlsx"

In my use cases any of the ways above would be helpful.

**Reply by noisyneil** [reply]
My personal preference would be for a command line interface. In combination with automation software like Keyboard Maestro or AutoHotkey, all sorts of situational triggers would be possible.

**Comment by roycetech** [comment]
I'm looking for this functionality as well. It is currently possible to use AppleScript to switch the profile via the menu, but it is very slow (~6s total to click the menu and then click the menu item). A faster way also via AppleScript is to launch the editor UI and then click via the profiles menu button. Still clunky, but it gets the job done. One nice solution I could think of would be a URL scheme. Something like streamdeck://switch=<profile_name>. I also wouldn't mind if they could add AppleScript support or simply make the app UI more accessible, and users can do all sorts of things with the app through AppleScript.

**Reply by noisyneil** [reply]
Have you tried the blank applet method (mentioned above)? It's much better than AS or GUI interaction. Faster and obtrusive.

**Reply by UweSchwo** [reply]
I'm not sure if I do understand what do you mean with "blank applet method". But I think having some exe Files only for the purpose of triggering profiles is to complicated cause you need front windows for triggering SD profiles.

**Reply by roycetech** [reply]
I'm trying the blank app today and it looks promising!

**Reply by noisyneil** [reply]
@UweSchwo, not sure I follow what you mean about needing front windows... Do you mean you want to trigger a profile for a particular front window? That's what I'm doing using blank apps. The apps handle the profile change itself but you have to use something like Auto Hotkey (or in my case Keyboard Maestro as I'm on Mac) to trigger opening them when a particular window is frontmost (or some other appropriate trigger event). 

(I called them applets because I'm on Mac, but yeah, they're just applications that do nothing except open and close, thereby triggering a SD profile change.)

**Reply by UweSchwo** [reply]
Sorry, I still do not understand it although I know AHK. 
Do you use the SD library inside AHK to trigger profiles by AHK code? 
Or do you just start any AHKtoExe generated exe file in order to trigger an automatic profile change (with any SD profile connected to this exe-file)? This doesn't work on my windows system. I must at least click on the running script Icon in systray  in order to bring it for a moment to foreground and thus trigger the automatic profile change.  Alternatively the AHK Script needs a GUI.

**Reply by noisyneil** [reply]
I can't speak for AHK because I'm on Mac, but I can imagine the same logic may apply. 

Keyboard Maestro has various triggers, one of which is _Focused Window Trigger_. This has a few options, like "The focused window title changed". 

You can then use an _If/Else_ condition to test the current window title or browser URL (among many other things). 

As an example, here's how I would switch to a dedicated SD profile whenever I search for something on Google:

Trigger on focused window title change
If front browser URL includes "google.com"
Then open Google.app
Else do nothing 

Google.app is an app made using Script Debugger or Script Editor that does nothing except open and then immediately close again. It is associated with a SD profile in the SD software and thus switches the SD profile.

Hopefully that clarifies how it works on Mac. I've seen other people discuss this technique using .exe files on Windows, so I'm confident it's possible. 

A quick Google returned this:

https://www.reddit.com/r/ElgatoGaming/comments/d6bwfr/streamdeck_how_do_i_switch_profiles_with_a/

**Reply by UweSchwo** [reply]
I got it. I allready have a self made AHK script that does any actions in dependence of current window title (comparable to Keyboard Maestro). But as seen in the linked information the desired exe (one per profile!) has to open a box (foreground window) for a moment in order to trigger the automatic profile change. Alternatively I can use some key combinations over Win-B.

Both methods are not solutions but workarounds you can lost the overview with. Especially if you often change the focused foreground window like I do. 
Meanwhile I use the standard profile changing to have at least my special buttons when Excel or Firefox are open, still independent of the open file / website. 

But thanks for focusing my attention to other AHK users with the same problem.

**Reply by roycetech** [reply]
I got mine to work as well. At first, I tried to prevent the app from appearing in the dock by modifying the LSUI attribute, but it didn't trigger the profile change. I changed it instead to a regular empty app (default options exported via Script Editor) and after briefly appearing in the dock, it switched the profile as expected. Better than my previous solution, thank you.

**Reply by noisyneil** [reply]
@UweSchwo FWIW, the apps don't need to open windows on Mac, but if they do on Windows, that's a bummer. 

@elgatosyElgato could surely implement some kind of URL scheme or shell script method of profile switching quite easily. After all, you can switch profiles from a second SD so there must be some kind of existing method in place. (What do I know though?!)

**Comment by markcellis** [comment]
I also would like the ability to dynamically switch to a user-created profile based on the profile name from my plugin. I tried to use the switchToProfile event but as the documentation notes: a plugin can only switch to read-only profiles declared in its manifest.json file.

**Comment by roycetech** [comment]
It's actually a lot faster now to switch profile via menu using AppleScript, on macOS.

```AppleScript
(* 
	@returns true if the profile was found.
*)
on switchProfile(deviceName, profileName)	
	tell application "System Events" to tell process "Stream Deck"
		try
			click menu item profileName of menu 1 of menu item deviceName of menu 1 of menu bar 2
			return true
		end try
	end tell			
	false
end switchProfile
```

**Reply by noisyneil** [reply]
> It's actually a lot faster now to switch profile via menu using AppleScript, on macOS.
> 
> ```applescript
> (* 
> 	@returns true if the profile was found.
> *)
> on switchProfile(deviceName, profileName)	
> 	tell application "System Events" to tell process "Stream Deck"
> 		try
> 			click menu item profileName of menu 1 of menu item deviceName of menu 1 of menu bar 2
> 			return true
> 		end try
> 	end tell			
> 	false
> end switchProfile
> ```

Holy moly! This even works if the menu bar item is hidden by Bartender!!! Might have to rethink my macro now! Here's a basic version that works nicely!

<img src="https://github.com/user-attachments/assets/0aa567ce-b605-4251-bfb8-936577d05186" width="65%" alt="Screenshot">

**Reply by noisyneil** [reply]
Ok I've added a fully-featured AppleScript version to the Keyboard Maestro [forum post](https://forum.keyboardmaestro.com/t/logic-pro-stream-deck-profile-switcher-2025-update/23357)!

**Comment by markcellis** [comment]
@ChekTek Why did you resolve this? To the best of my knowledge there is still no way to switch the profile via the SDK. If there is please point me to it.

**Reply by ChekTek** [reply]
Hi @markcellis, 

You can switch to profiles bundled within the plugin: https://docs.elgato.com/streamdeck/sdk/guides/profiles

We have also added the request to switch to user created profiles to the Stream Deck application backlog as this was not the first request for this feature.

**Reply by noisyneil** [reply]
I have to agree that it would be useful to keep this topic open so that users can continue to share improved workarounds until that hallowed day when programmatic profile switching is finally implemented. 

I'm glad to hear it's been added to the "backlog", and I hope it's high up on the list, because it would exponentially increase the usefulness of the device. At the very least, switches based on front window title matches would be very useful, even for those users who don't use secondary automation tools like Keyboard Maestro and the like.

**Reply by ChekTek** [reply]
With no discussion happening in nearly a full year, I assumed this discussion was stale. Apologies it's seems like this question has not been answered in this thread. We've stated on multiple occasions that the team is aware of this particular request, but not in this thread. I'll re-open the issue for now.

**Reply by markcellis** [reply]
I'm glad it is on the backlog. My users want my plugin to be able to switch between profiles they created.

**Reply by UweSchwo** [reply]
Yes, it's important to put this problem in focus again. The automatic profile switching for me is not very stable and not granular enough. I'd like to replace the whole mechanism by some ahk triggered commands.

**Reply by noisyneil** [reply]
I'd just like to let the Keyboard Maestro users among us know that I recently updated the macro in the initial post. It's now _much_ simpler to set up and includes a tool that automatically creates the required trigger apps that trigger profiles for Logic Pro plugins. If there's enough interest, I'd be happy to adapt this for more general use (websites etc).
